go-controller/pkg/util/timeoutqueue/timequeue.go

```diff
@@ -0,0 +1,210 @@
+package timeoutqueue
+
+import (
+	"container/heap"
+	"context"
+	"sync"
+	"time"
+)
+
+// TimeoutQueue is a thread safe queue of items that can only be popped after a
+// timeout
+type TimeoutQueue[T comparable] struct {
+	popM      sync.Mutex
+	pushM     sync.Mutex
+	items     heapImpl[T]
+	consumers map[chan struct{}]time.Time
```

Comment:
So consumers are callers of `pop` and producers are callers of `push` and the whole thing is thread safe. 

Consumers register themselves to the consumers map with a channel and the item deadline which they will be blocked waiting on.

Producers look in that map for a consumer waiting on a item with a deadline later than the items that's being pushed. If found, we notify that consumer through the corresponding channel so that the consumer requeues the item that it was waiting on and picks a new item to wait on, which should be one with an earlier deadline if available.

So the channel is used for a producer to notify a consumer that a best candidate item is available to consume than the one it was holding on.

Reply:
I still don't understand, but my instincts tell me this seems like overkill. I can see why you want to use a "TimeoutQueue", but I would just register a single channel that items are returned on (not a map). 

Then, I would not have the caller/consumer running in infinite for loop trying to pop items. I would expect the TimeoutQueue would send me back the item on the channel once the time is up and it should run the for loop.